### PR TITLE
#1019 Migrated LPUSH, RPUSH, LPOP, RPOP, LLEN

### LINKED ISSUE #1019: Command Migration: ('LPUSH', 'RPUSH', 'LPOP', 'RPOP', 'LLEN')
This issue tracks the migration of the mentioned commands - ('LPUSH', 'RPUSH', 'LPOP', 'RPOP', 'LLEN') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by Aditya-Chowdhary**
fixes #1019 

Edit: Now ready for review
Currently draft PR as updating integration tests is remaining. Remaining tasks are completed.

For migration commands, please consider this a template on how to tackle the issue. I have divided the tasks into smaller goals and made an individual commit.
| Condition                                      | Return Value                                      |
|------------------------------------------------|---------------------------------------------------|
| Migrating eval to store eval | 
|Modify unit test with new response format |
| Migrate integration tests from async to resp|
| Add/update integration in websocket |
| Add/update integration in HTTP |
| Ensure all integration test pass |
| Update/Add docs for the commands |
| Lint checks | |

- [x]  Migrated the evalXXX function with the latest definition
- [x]  Update or add unit tests for the new implementation.
- [x]  All unit tests pass successfully.
- [x]  Ensure all integration tests pass successfully (Resp/http/websocket)
- [x] Move Integration tests for the respective commands under the RESP integration tests directory from the Async directory
- [x] Please validate that the documentation for the respective commands is up to date. If not then consider adding them.

### CODE REVIEW

File: internal/eval/store_eval.go [hunk 1]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	}
+
+	obj := store.Get(args[0])
+	if obj == nil {
+		obj = store.NewObj(NewDeque(), -1, object.ObjTypeByteList, object.ObjEncodingDeque)
+	}
+
+	if err := object.AssertType(obj.TypeEncoding, object.ObjTypeByteList); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2064]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2064]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 2]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongTypeOperation,
+		}
+	}
+
+	if err := object.AssertEncoding(obj.TypeEncoding, object.ObjEncodingDeque); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2071]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2071]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
 		Error:  nil,
 	}
 }
+
+func evalLPUSH(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) < 2 {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2052]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2052]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 4]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	return &EvalResponse{
+		Result: deq.Length,
+		Error:  nil,
+	}
+}
+
+func evalRPUSH(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) < 2 {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2092]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2092]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 5]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	}
+
+	obj := store.Get(args[0])
+	if obj == nil {
+		obj = store.NewObj(NewDeque(), -1, object.ObjTypeByteList, object.ObjEncodingDeque)
+	}
+
+	if err := object.AssertType(obj.TypeEncoding, object.ObjTypeByteList); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2104]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2104]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 6]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongTypeOperation,
+		}
+	}
+
+	if err := object.AssertEncoding(obj.TypeEncoding, object.ObjEncodingDeque); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2111]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2111]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 7]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	return &EvalResponse{
+		Result: deq.Length,
+		Error:  nil,
+	}
+}
+
+func evalLPOP(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) != 1 {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2132]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2132]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 8]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongArgumentCount("LPOP"),
+		}
+	}
+
+	obj := store.Get(args[0])
+	if obj == nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2140]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2140]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 9]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  nil,
+		}
+	}
+
+	if err := object.AssertType(obj.TypeEncoding, object.ObjTypeByteList); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2147]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2147]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 10]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongTypeOperation,
+		}
+	}
+
+	if err := object.AssertEncoding(obj.TypeEncoding, object.ObjEncodingDeque); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2154]:
```suggestion
			Result: nil,
```

Comment (AshwinKul28) [reply internal/eval/store_eval.go:2154]: Please send nil, when sending Error from eval.

File: internal/eval/store_eval.go [hunk 11]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	deq := obj.Value.(*Deque)
+	x, err := deq.LPop()
+	if err != nil {
+		if errors.Is(err, ErrDequeEmpty) {
+			return &EvalResponse{
+				Result: clientio.NIL,
+				Error:  nil,
+			}
+		}
+		panic(fmt.Sprintf("unknown error: %v", err))
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2168]: Any idea why we are panicking here? You can send Error as Internal Server but we shouldn't panic the server.

File: internal/eval/store_eval.go [hunk 12]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	return &EvalResponse{
+		Result: x,
+		Error:  nil,
+	}
+}
+
+func evalRPOP(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) != 1 {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2180]:
```suggestion
			Result: nil,
```

File: internal/eval/store_eval.go [hunk 13]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongArgumentCount("RPOP"),
+		}
+	}
+
+	obj := store.Get(args[0])
+	if obj == nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2188]:
```suggestion
			Result: nil,
```

File: internal/eval/store_eval.go [hunk 14]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  nil,
+		}
+	}
+
+	if err := object.AssertType(obj.TypeEncoding, object.ObjTypeByteList); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2195]:
```suggestion
			Result: nil,
```

File: internal/eval/store_eval.go [hunk 15]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongTypeOperation,
+		}
+	}
+
+	if err := object.AssertEncoding(obj.TypeEncoding, object.ObjEncodingDeque); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2202]:
```suggestion
			Result: nil,
```

File: internal/eval/store_eval.go [hunk 16]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  diceerrors.ErrWrongTypeOperation,
+		}
+	}
+
+	deq := obj.Value.(*Deque)
+	x, err := deq.RPop()
+	if err != nil {
+		if errors.Is(err, ErrDequeEmpty) {
+			return &EvalResponse{
+				Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2212]:
```suggestion
				Result: nil,
```

File: internal/eval/store_eval.go [hunk 17]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	return &EvalResponse{
+		Result: x,
+		Error:  nil,
+	}
+}
+
+func evalLLEN(args []string, store *dstore.Store) *EvalResponse {
+	if len(args) != 1 {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2228]:
```suggestion
			Result: nil,
```

File: internal/eval/store_eval.go [hunk 18]
```diff
@@ -2044,3 +2045,209 @@ func evalZPOPMAX(args []string, store *dstore.Store) *EvalResponse {
...
+	if obj == nil {
+		return &EvalResponse{
+			Result: clientio.IntegerZero,
+			Error:  nil,
+		}
+	}
+
+	if err := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeByteList, object.ObjEncodingDeque); err != nil {
+		return &EvalResponse{
+			Result: clientio.NIL,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:2243]:
```suggestion
			Result: nil,
```

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]
HI @Aditya-Chowdhary Thanks a lot for these changes. I have requested changes mostly in return responses, where while sending error we shouldn't send Result as clientio.NIL, instead it should be just nil.

File: internal/eval/store_eval.go (hunk 8 above)
Comment (Aditya-Chowdhary) [reply internal/eval/store_eval.go:2140]: This one does not send error from eval, so I have left this as clientio.NIL

File: internal/eval/store_eval.go (hunk 13 above)
Comment (Aditya-Chowdhary) [reply internal/eval/store_eval.go:2188]: Similar to above, as we aren't returning an error, I have left this clientio.Nil

File: internal/eval/store_eval.go (hunk 16 above)
Comment (Aditya-Chowdhary) [reply internal/eval/store_eval.go:2212]: Similar to above, as we do not return an error, I have left this clientio.NIL

File: internal/eval/store_eval.go (hunk 11 above)
Comment (Aditya-Chowdhary) [reply internal/eval/store_eval.go:2168]: You're right, I looked further into the function, there's no need for the panic, so I have removed it.

**Comment by Aditya-Chowdhary** [comment]
@AshwinKul28 I have made the requested changes. Several of the ones you commented on did not return an error, so I have left those as clientio.NIL for now, please let me know if I should change those ones as well. 

https://github.com/DiceDB/dice/pull/1181/files#r1818149611
https://github.com/DiceDB/dice/pull/1181/files#r1818150264
https://github.com/DiceDB/dice/pull/1181/files#r1818150407

File: internal/eval/store_eval.go (hunk 8 above)
Comment (AshwinKul28) [reply internal/eval/store_eval.go:2140]: Alright, that sounds good!

File: internal/eval/store_eval.go (hunk 13 above)
Comment (AshwinKul28) [reply internal/eval/store_eval.go:2187]: sounds good!

File: internal/eval/store_eval.go (hunk 16 above)
Comment (AshwinKul28) [reply internal/eval/store_eval.go:2211]: Looks great!

**Comment by AshwinKul28** [comment]
HI @Aditya-Chowdhary thanks again for the commendable efforts. The changes look great. 

Please rebase your branch with the master and I have added a checklist in the PR description, go through it and complete those tasks. (Most of them already you have taken care :D)

For the documentation of LPUSH, RPUSH, RPOP commands we need a refresher based on this [sample doc](https://github.com/DiceDB/dice/blob/master/docs/sample_command_docs.md) such as example descriptions, notes and best practices if required.

**Comment by JyotinderSingh** [comment]
There is an integration test failure, could you please take a look?

**Comment by Aditya-Chowdhary** [comment]
@JyotinderSingh Have fixed any merge conflicts + test failures

File: config/config.go
Comment (apoorvyadav1111) [inline config/config.go:1]: Please revert this change.

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
Just a small change and we are good to merge

File: config/config.go [hunk 19]
```diff
@@ -27,7 +27,7 @@ const (
 	EvictAllKeysLRU    = "allkeys-lru"
 	EvictAllKeysLFU    = "allkeys-lfu"
 
-	DefaultKeysLimit int = 200000000
+	DefaultKeysLimit int = 2000
```
Comment (apoorvyadav1111) [inline config/config.go:30]:
```suggestion
	DefaultKeysLimit int = 200000000
```

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]

**Review by apoorvyadav1111 — APPROVED** [review]

**Comment by apoorvyadav1111** [comment]
Thanks for taking up this issue and closing it.
